urllib3/util/hpkp.py

```diff
@@ -0,0 +1,269 @@
+            part_count = len(split_host)
+
+            if part_count >= split_domain_part_count:
+                # Too many parts, cannot be a parent domain.
+                continue
+
+            if split_host == split_domain[:part_count]:
+                # Parent domain!
+                hosts.append(host)
+
+        # Sort by domain length, longest first.
+        hosts.sort(key=lambda h: len(h.domain), reverse=True)
+
+        return hosts
+
+
+class EphemeralHPKPDatabase(HPKPDatabase):
+    """
+    The default, in-memory HPKP database. This provides ephemeral HPKP storage,
+    in-memory. It is not recommended for high-security production use, but it
+    is better than nothing.
+    """
+    def __init__(self):
+        self.hosts = {}
+
+    def store_host(self, known_host):
+        self.hosts[known_host.domain] = known_host
+
+    def invalidate_host(self, domain):
+        del self.hosts[domain]
+
+    def iter_hosts(self):
+        return self.hosts.values()
+
+
+class HPKPManager(object):
+    """
+    An object that manages HPKP for urllib3.
+
+    This object is responsible for validating HPKP for HTTPS connections. It
+    employs a :class:`HPKPDatabase <urllib3.util.hpkp.HPKPDatabase>` object to
+    manage the actual storage, but the logic is contained in this class.
+    """
+    def __init__(self, database):
+        self.db = database
+
+    def validate_connection(self, domain, socket):
+        """
+        For a given connection, confirm that it passes HPKP validation.
+
+        This checks whether there are any ``KnownPinnedHost``s for the given
+        connection. If they are, validates that the presented certificate
+        actually matches the pin.
+        """
+        hosts = self.db.get_hosts_by_domain(domain)
+        if not hosts:
+            # This connection has not previously done any key pinning, so
+            # there's no validation to do.
+            return
+
+        host = self._find_valid_pinned_host(domain, hosts)
+        if host is None:
+            # Again, no previous pinning was done on this connection, so
+            # there's no validation to do.
+            return
+
+        # If we got this far, we have a KnownPinnedHost in hand that applies to
+        # this connection. Get the certificate and check it matches one of the
+        # pins.
+        # TODO: Actually do this!
+        return
+
+    def process_response(self, domain, response):
+        """
+        Processes a HTTPS response for HPKP.
+
+        This method checks whether the PKP header is present on the request,
+        and if it is updates its store of key pins. It also validates the
+        connection against the HPKP header, confirming that it does pass at
+        this stage.
+        """
+        # TODO: Handle ip address domains.
+        # If no PKP header is provided, then no action is required.
+        pkp = response.getheader('public-key-pins')
```

Comment:
Fair enough, it's a big diff (already), and it might be the case that we want to move some of this logic into these classes anyway.